hello-compose-app/src/main/kotlin/hello/com/pose/MainScreen.kt

```diff
@@ -15,42 +13,82 @@ import androidx.compose.ui.unit.dp
 import androidx.lifecycle.viewmodel.compose.viewModel
 import coil.ImageLoader
 import coil.compose.rememberAsyncImagePainter
+import hello.com.pose.composable.Alert
 import hello.com.pose.composable.SearchBar
 import hello.com.pose.shared.domain.photo.Photo
+import kotlinx.coroutines.launch
 
 @Composable
 fun MainScreen(viewModel: MainViewModel = viewModel()) {
-    val query: MutableState<String> = remember { mutableStateOf("") }
+    val lazyGridState = rememberLazyGridState()
+    val coroutineScope = rememberCoroutineScope()
+    val currentSearchQuery by viewModel.searchQuery.collectAsState()
+
+    val shouldStartPaginate = remember {
+        derivedStateOf {
+            viewModel.canPaginate && lazyGridState.isLastItemVisible
+        }
+    }
+
+    LaunchedEffect(key1 = shouldStartPaginate.value) {
+        if (shouldStartPaginate.value && viewModel.pagingState == PagingState.IDLE) {
+            viewModel.getPhotosByQuery(currentSearchQuery)
+        }
+    }
 
     Surface(modifier = Modifier.fillMaxSize()) {
         Column(modifier = Modifier.padding(8.dp)) {
-            SearchBar(text = query.value,
+            SearchBar(text = currentSearchQuery,
                 inputChange = {
-                    query.value = it
-                    viewModel.getImageList(query.value)
+                    viewModel.setNewQuery(it)
                 })
-
-            LazyVerticalGrid(columns = GridCells.Fixed(4)) {
-                viewModel.photos.value.data.let { photo ->
-                    items(photo) {
-                        MainContentItem(photo = it)
+            SearchList(viewModel, lazyGridState)
+            LaunchedEffect(key1 = currentSearchQuery) {
+                if (viewModel.prevQuery != currentSearchQuery) {
+                    coroutineScope.launch {
+                        lazyGridState.animateScrollToItem(index = 0)
                     }
                 }
             }
         }
     }
 }
 
+@Composable
+fun SearchList(viewModel: MainViewModel, gridState: LazyGridState) {
+    LazyVerticalGrid(
+        columns = GridCells.Fixed(4),
+        state = gridState
+    ) {
+        items(viewModel.photoList) { photo ->
+            MainContentItem(photo = photo)
+        }
+    }
+}
+
+val LazyGridState.isLastItemVisible: Boolean
+    get() = layoutInfo.visibleItemsInfo.lastOrNull()?.index == layoutInfo.totalItemsCount - 1
+
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
 fun MainContentItem(photo: Photo) {
+    val showDownloadDialog = remember { mutableStateOf(false) }
+
+    if (showDownloadDialog.value) {
+        Alert(
+            showDialog = showDownloadDialog.value,
+            onDismiss = { showDownloadDialog.value = false },
+            photo
+        )
+    }
+
     Card(
         modifier = Modifier
             .padding(2.dp)
```

Comment:
카드의 간격이 필요하다면 padding없이 LazyVerticalGrid의 contentPadding를 이용하면됩니다!

Reply:
아하 contentPadding으로 수정하겠습니다!

Reply:
```suggestion
            LazyVerticalGrid(
                  contentPadding = PaddingValues(all = 10.dp)
    )
```
위와 같이 contentPadding을 적용하면 카드의 간격에 패딩이 먹는게 아니라 리스트에 먹히는거 같은데 이렇게 적용하는게 아닐까요?

<img width="250" alt="스크린샷 2023-02-15 오후 2 01 43" src="https://user-images.githubusercontent.com/40010002/218936364-b0b5572c-eb1b-4b9e-94bd-efcef7e4f942.png">

Reply:
에고 제가 API를 헷갈렸네요 ㅠ ..
verticalArrangement = Arrangement.spacedBy(2.dp)
horizontalArrangement = Arrangement.spacedBy(2.dp)
그리드 마진값을 구성할 때 이렇게 구현했었어요!